apps/remote_control/lib/lexical/remote_control/search/indexer/source/reducer.ex

```diff
@@ -0,0 +1,145 @@
+defmodule Lexical.RemoteControl.Search.Indexer.Source.Reducer do
+  @moduledoc """
+  A module and struct that can reduce over elixir AST via Macro.prewalk/3
+
+  The reducer keeps track of blocks and parent / child relationships so extractors don't have to concern themselves
+  with the AST's overall structure, and can focus on extracting content from it.
+  """
+  alias Lexical.Document
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+  alias Lexical.RemoteControl.Search.Indexer.Extractors
+  alias Lexical.RemoteControl.Search.Indexer.Metadata
+  alias Lexical.RemoteControl.Search.Indexer.Source.Block
+
+  defstruct [:entries, :document, :quoted_document, :position, :ends_at, :blocks]
+
+  @extractors [Extractors.Module]
+
+  def new(%Document{} = document, quoted_document) do
+    %__MODULE__{
+      document: document,
+      quoted_document: quoted_document,
+      entries: [],
+      position: {0, 0},
+      blocks: [Block.root()]
+    }
+  end
+
+  def entries(%__MODULE__{} = reducer) do
+    Enum.reverse(reducer.entries)
+  end
+
+  def skip(meta) do
+    Keyword.put(meta, :__skipped__, true)
+  end
+
+  def skip?({_, meta, _}) do
+    Keyword.get(meta, :__skipped__, false)
+  end
+
+  def skip?(_), do: false
+
+  def reduce(%__MODULE__{} = reducer, element) do
+    if skip?(element) do
+      {reducer, element}
+    else
+      do_reduce(reducer, element)
+    end
+  end
+
+  defp do_reduce(%__MODULE__{} = reducer, element) do
+    case Metadata.location(element) do
+      {:block, position, block_start, block_end} ->
+        block = Block.new(block_start, block_end)
+
+        reducer
+        |> update_position(position)
+        |> maybe_pop_block()
+        |> push_block(block)
```

Comment:
I haven't fully read through this PR, but I feel that the core should be this module. Here, I don't quite understand why we initially pop a block and then push it back (with a parent ref added).

I would like to know the reason behind this. Could you provide an example to help me better understand it?

Reply:
Sure. This call happens when we encounter a new block. We're _maybe_ popping the existing block, and then pushing the new block. This allows us to differentiate between:
```elixir
defmodule First do 
end
defmodule Second do <position> 
end
```
and 
```elixir
defmodule Parent do 
  defmodule Child do <position>
  end 
end
```

In the first, `maybe_pop_block` pops the block representing the `First` module, and in the second, `maybe_pop_block` doesn't do anything, keeping the parent block in-scope and child block is pushed.